signingscript/sign.py

```diff
@@ -625,3 +637,64 @@ def _get_tarfile_compression(compression):
         return to
     except Exception as e:
         raise SigningScriptError(e)
+
+
+async def sign_file_with_autograph(context, from_, fmt, to=None):
+    """Signs a file with autograph and writes the result to arg `to` or `from_` if `to` is None.
+
+    Args:
+        context (Context): the signing context
+        from_ (str): the source file to sign
+        fmt (str): the format to sign with
+        to (str, optional): the target path to sign to. If None, overwrite
+            `from_`. Defaults to None.
+
+    Raises:
+        Requests.RequestException: on failure
+
+    Returns:
+        str: the path to the signed file
+
+    """
+    if not utils.is_autograph_signing_format(fmt):
+        raise SigningScriptError(
+            "No signing servers found supporting signing format {}".format(fmt)
+        )
+    cert_type = task.task_cert_type(context)
+    servers = get_suitable_signing_servers(context.signing_servers, cert_type, [fmt])
+    if not servers:
+        raise SigningScriptError(
+            "No signing servers found with cert type {}".format(cert_type)
+        )
+    s = servers[0]
+    to = to or from_
+
+    with open(from_, 'rb') as fin:
+        input_bytes = fin.read()
+
+    # build and run the signature request
+    sign_req = [{"input": base64.b64encode(input_bytes)}]
+    log.debug("using the default autograph keyid for %s", s.user)
+    log.debug("autograph signature request: %s" % sign_req)
+
+    url = "%s/sign/file" % s.server
+    auth = HawkAuth(id=s.user, key=s.password)
+    retries = Retry(**task.get_autograph_retry_config(context))
+
+    def make_sign_req():
+        with requests.Session() as session:
+            session.mount('http://', HTTPAdapter(max_retries=retries))
+            session.mount('https://', HTTPAdapter(max_retries=retries))
+            return session.post(url, json=sign_req, auth=auth)
+
+    loop = asyncio.get_event_loop()
+    response = await loop.run_in_executor(None, make_sign_req)
+    response.raise_for_status()
+    log.debug("autograph signature response: %s" % response.text)
+    sign_resp = response.json()
```

Comment:
We can avoid buffering the whole response with requests' [stream=True](http://docs.python-requests.org/en/latest/user/advanced/#body-content-workflow) and https://docs.python.org/3/library/base64.html#base64.decode

Reply:
to get streaming working I think we need the following:

1. write the request body as a stream (it should look like `b'[{"input":"'<base64 encoded chunks of the file>"}]'`)
1. [disable including the body/content in the hash MAC](https://mohawk.readthedocs.io/en/latest/usage.html#skipping-content-checks) or update requests_hawk and mohawk to handle streams
1. read the response body as a stream
1. read JSON until we find `"signed_file":"` in the response body
1. base64 decode chunks of the signed file
1. write chunks to disk
1. stop at `"`

I tried:

* with requests, use [stream=True](http://docs.python-requests.org/en/latest/user/advanced/#body-content-workflow) and [response.iter_content](http://docs.python-requests.org/en/master/api/#requests.Response.iter_content)

* with aiofile and aiohttp, reuse `context.session`, post with [an async file_sender generator](https://aiohttp.readthedocs.io/en/stable/client_quickstart.html#streaming-uploads) in the data param and read [the response.content stream](https://aiohttp.readthedocs.io/en/stable/client_reference.html#response-object) since aiohttp.StreamReader implements asyncio.StreamReader so we can use [readuntil](https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamReader.readuntil) which is nice. It does require switching backoff/retry libraries to something like https://github.com/litl/backoff/

I ran into problems padding errors streaming base64 encode/decode (since it's mapping 6 bits to 8 bits) and using the dev.mozaws.net cert with aiohttp (pulled it down with openssl and put it in cert path, but still got cert validation errors) and anticipate problems checking for `signed_file` occurring at a chunk boundary, disabling the hawk content hashing being not great.

We could also change the autograph API to not use hawk or support chunked HTTP requests and responses, but I assume that's not really an option.

If the mar data to sign is small enough to buffer in memory and send to autograph's /sign/data then that should maximize throughput, minimize network overhead, and yield more robust and maintainable code. We'd probably need to do something like monkeypatch make_signer_v2 in [mardor](https://github.com/mozilla/build-mar/) too.

Reply:
@catlee do you have concerns about landing as-is, given we're signing a file at a time? (Is this a blocker to land, or a followup task?)

Reply:
I think we can probably test with it as-is, and deal with it if it's a problem.